cpp/src/arrow/util/logging.h

```diff
@@ -81,23 +81,36 @@ enum class ArrowLogLevel : int {
 #ifdef NDEBUG
 #define ARROW_DFATAL ::arrow::util::ArrowLogLevel::ARROW_WARNING
 
-#define DCHECK(condition) \
-  while (false) ::arrow::util::detail::NullLog()
-#define DCHECK_OK(s)    \
-  ARROW_IGNORE_EXPR(s); \
-  while (false) ::arrow::util::detail::NullLog()
-#define DCHECK_EQ(val1, val2) \
-  while (false) ::arrow::util::detail::NullLog()
-#define DCHECK_NE(val1, val2) \
-  while (false) ::arrow::util::detail::NullLog()
-#define DCHECK_LE(val1, val2) \
-  while (false) ::arrow::util::detail::NullLog()
-#define DCHECK_LT(val1, val2) \
-  while (false) ::arrow::util::detail::NullLog()
-#define DCHECK_GE(val1, val2) \
-  while (false) ::arrow::util::detail::NullLog()
-#define DCHECK_GT(val1, val2) \
-  while (false) ::arrow::util::detail::NullLog()
+#define DCHECK(condition)                     \
+  while (false) ARROW_IGNORE_EXPR(condition); \
+  while (false) ::arrow::util::Voidify() & ::arrow::util::detail::NullLog()
+#define DCHECK_OK(s)                  \
+  while (false) ARROW_IGNORE_EXPR(s); \
```

Comment:
Previously `s` was computed (see above). This is probably the source of the segfaults on some configurations (see e.g. AppVeyor results).

Reply:
It's confusing that `DCHECK_OK` always computes its expression but not the other `DCHECK` macros, we might remove `DCHECK_OK` and replace it with `ABORT_NOT_OK`.

Reply:
Ah, I forgot DCHECK_OK needs to compute the expression. Thanks